File: lib-impl/src/main/java/org/ergoplatform/appkit/impl/ErgoProverImpl.java
package org.ergoplatform.appkit.impl;

import org.ergoplatform.ErgoBox;
import org.ergoplatform.ErgoLikeTransaction;
import org.ergoplatform.P2PKAddress;
import org.ergoplatform.appkit.*;
import org.ergoplatform.wallet.interpreter.ErgoProvingInterpreter;
import scala.collection.IndexedSeq;
import sigmastate.basics.DLogProtocol;

public class ErgoProverImpl implements ErgoProver {
    private final BlockchainContextImpl _ctx;
    private final ErgoProvingInterpreter _prover;

    public ErgoProverImpl(BlockchainContextImpl ctx, ErgoProvingInterpreter prover) {
        _ctx = ctx;
        _prover = prover;
    }

    @Override
    public P2PKAddress getP2PKAddress() {
        DLogProtocol.ProveDlog pk = _prover.secrets().apply(0).publicImage();
        return JavaHelpers.createP2PKAddress(pk, _ctx.getNetworkType().networkPrefix);
    }

    @Override
    public Address getAddress() {
       return new Address(getP2PKAddress());
    }

    @Override
    public SignedTransaction sign(UnsignedTransaction tx) {
        UnsignedTransactionImpl txImpl = (UnsignedTransactionImpl)tx;
        IndexedSeq<ErgoBox> boxesToSpend = JavaHelpers.toIndexedSeq(txImpl.getBoxesToSpend());
        IndexedSeq<ErgoBox> dataBoxes = JavaHelpers.toIndexedSeq(txImpl.getDataBoxes());
        ErgoLikeTransaction signed =
                _prover.sign(txImpl.getTx(), boxesToSpend, dataBoxes, txImpl.getStateContext()).get();
        return new SignedTransactionImpl(_ctx, signed);
    }
}

File: lib-impl/src/main/java/org/ergoplatform/appkit/impl/ErgoProverImpl.scala
package org.ergoplatform.appkit.impl

import org.ergoplatform.appkit._
import org.ergoplatform.sdk.wallet.secrets.ExtendedSecretKey
import org.ergoplatform.sdk.{AppkitProvingInterpreter, JavaHelpers, UnreducedTransaction}
import org.ergoplatform.{P2PKAddress, sdk}
import sigma.data.CSigmaDslBuilder
import sigmastate.interpreter.HintsBag
import sigma.BigInt

import java.util
import scala.collection.JavaConverters._

class ErgoProverImpl(_ctx: BlockchainContextBase,
                     _prover: AppkitProvingInterpreter) extends ErgoProver {
  private def networkPrefix = _ctx.getNetworkType.networkPrefix

  override def getP2PKAddress: P2PKAddress = {
    val pk = _prover.pubKeys(0)
    JavaHelpers.createP2PKAddress(pk, networkPrefix)
  }

  override def getAddress = new Address(getP2PKAddress)

  override def getSecretKey: BigInt =
    CSigmaDslBuilder.BigInt(_prover.secretKeys(0).privateInput.w)

  override def getEip3Addresses: util.List[Address] = {
    val addresses = _prover.secretKeys.toIndexedSeq
      .drop(1)
      .map { k =>
        val p2pkAddress = JavaHelpers.createP2PKAddress(k.publicImage, networkPrefix)
        new Address(p2pkAddress)
      }
    addresses.asJava
  }

  override def sign(tx: UnsignedTransaction): SignedTransaction =
    sign(tx, baseCost = 0)

  override def sign(tx: UnsignedTransaction, baseCost: Int): SignedTransaction = {
    val txImpl = tx.asInstanceOf[UnsignedTransactionImpl]
    val boxesToSpend = JavaHelpers.toIndexedSeq(txImpl.getBoxesToSpend)
    val dataBoxes = JavaHelpers.toIndexedSeq(txImpl.getDataBoxes)
    val tokensToBurn = JavaHelpers.toIndexedSeq(txImpl.getTokensToBurn)
    val unreduced = UnreducedTransaction(txImpl.getTx, boxesToSpend, dataBoxes, tokensToBurn)
    val signed = _prover.sign(
      unreducedTx = unreduced,
      stateContext = txImpl.getStateContext,
      baseCost = baseCost).get
    new SignedTransactionImpl(_ctx, signed.ergoTx, signed.cost)
  }

  override def signMessage(sigmaProp: SigmaProp, message:  Array[Byte], hintsBag: HintsBag): Array[Byte] = {
    _prover.signMessage(sigmaProp.getSigmaBoolean, message, hintsBag).get
  }

  override def reduce(tx: UnsignedTransaction, baseCost: Int): ReducedTransaction = {
    val txImpl = tx.asInstanceOf[UnsignedTransactionImpl]
    val boxesToSpend = JavaHelpers.toIndexedSeq(txImpl.getBoxesToSpend)
    val dataBoxes = JavaHelpers.toIndexedSeq(txImpl.getDataBoxes)
    val tokensToBurn = JavaHelpers.toIndexedSeq(txImpl.getTokensToBurn)
    val reduced = _prover.reduceTransaction(
      unsignedTx = txImpl.getTx,
      boxesToSpend = boxesToSpend,
      dataBoxes = dataBoxes,
      stateContext = txImpl.getStateContext,
      baseCost = baseCost,
      tokensToBurn = tokensToBurn)
    new ReducedTransactionImpl(_ctx, reduced)
  }

  override def signReduced(tx: ReducedTransaction, baseCost: Int): SignedTransaction = {
    val signed = _prover.signReduced(sdk.ReducedTransaction(tx.getTx), baseCost)
    new SignedTransactionImpl(_ctx, signed.ergoTx, signed.cost)
  }

}

